R/extract_consultations_erprsf.R

```diff
@@ -0,0 +1,272 @@
+    dis_dtd_lag_months = 6,
+    patients_ids = NULL,
+    output_table_name = NULL,
+    overwrite = FALSE,
+    conn = NULL) {
+  stopifnot(
+    !is.null(start_date),
+    !is.null(end_date),
+    inherits(start_date, "Date"),
+    inherits(end_date, "Date"),
+    start_date <= end_date
+  )
+  connection_opened <- FALSE
+  if (is.null(conn)) {
+    conn <- connect_oracle()
+    connection_opened <- TRUE
+  }
+  timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
+  if (!is.null(output_table_name)) {
+    output_table_name_is_temp <- FALSE
+    stopifnot(
+      is.character(output_table_name),
+      !DBI::dbExistsTable(conn, output_table_name) || (
+        DBI::dbExistsTable(conn, output_table_name) && overwrite
+      )
+    )
+    if (DBI::dbExistsTable(conn, output_table_name) && overwrite) {
+      warning(
+        glue::glue(
+          "Table {output_table_name} already exists and will be overwritten."
+        )
+      )
+      DBI::dbRemoveTable(conn, output_table_name)
+    }
+  } else {
+    output_table_name_is_temp <- TRUE
+    output_table_name <- glue::glue("TMP_DISP_{timestamp}")
+  }
+  if (!is.null(patients_ids)) {
+    stopifnot(
+      identical(
+        names(patients_ids),
+        c("BEN_IDT_ANO", "BEN_NIR_PSA")
+      ),
+      !anyDuplicated(patients_ids)
+    )
+    patients_ids_table_name <- glue::glue("TMP_PATIENTS_IDS_{timestamp}")
+    DBI::dbWriteTable(conn, patients_ids_table_name, patients_ids)
+  }
+
+  dis_dtd_end_date <-
+    end_date |>
+    lubridate::add_with_rollback(months(dis_dtd_lag_months)) |>
+    lubridate::floor_date("months")
+
+  start_year <- lubridate::year(start_date)
+  end_year <- lubridate::year(dis_dtd_end_date)
+
+  formatted_start_date <- format(start_date, "%Y-%m-%d")
+  formatted_end_date <- format(end_date, "%Y-%m-%d")
+  formatted_dis_dtd_end_date <- format(dis_dtd_end_date, "%Y-%m-%d")
+
+  first_non_archived_year <- get_first_non_archived_year(conn)
+
+  if (!is.null(pse_spe_codes)) {
+    print(
+      glue::glue(
+        "Extracting consultations from all specialties among  {paste(pse_spe_codes, collapse = ' or ')}..."
+      )
+    )
+  } else {
+    print(
+      glue::glue(
+        "Extracting consultations from all specialties codes..."
+      )
+    )
+  }
+
+  pb <- progress::progress_bar$new(
+    format = "Extracting :year1 (going from :year2 to :year3) [:bar] :percent in :elapsed (eta: :eta)",
+    total = (end_year - start_year + 1), clear = FALSE, width = 80
+  )
+  pb$tick(0)
+  for (year in start_year:end_year) {
+    pb$tick(tokens = list(year1 = year, year2 = start_year, year3 = end_year))
+    if (year < first_non_archived_year) {
+      er_prs_f <- dplyr::tbl(conn, glue::glue("ER_PRS_F_{year}"))
+    } else {
+      er_prs_f <- dplyr::tbl(conn, "ER_PRS_F")
+    }
+
+    if (year == end_year) {
+      dis_dtd_condition <- glue::glue(
+        "FLX_DIS_DTD >= DATE '{year}-02-01'
+        AND FLX_DIS_DTD <= DATE '{formatted_dis_dtd_end_date}'"
```

Comment:
pour toutes les conditions de la forme `x >= y AND x <= z`, on peut utiliser la forme moins verbeuse `x BETWEEN y AND z`